src/irmin-pack/unix/gc.ml

```diff
@@ -17,6 +17,310 @@
+       value. *)
+    let stats =
+      S.
+        {
+          stats with
+          steps = [];
+          commit_offset;
+          before_suffix_start_offset;
+          before_suffix_end_offset;
+          after_suffix_start_offset;
+        }
+    in
+    let timer = Steps_timer.create first_stepname in
+    { stats; timer }
+
+  let finish_current_step t next_stepname =
+    let timer, prev_step = Steps_timer.progress t.timer next_stepname in
+    let stats = { t.stats with steps = prev_step :: t.stats.steps } in
+    { stats; timer }
+
+  let finalise t worker ~after_suffix_end_offset =
+    let t = finish_current_step t "will not appear in the stats" in
+    {
+      t.stats with
+      S.worker;
+      after_suffix_end_offset;
+      steps = List.rev t.stats.steps;
+      suffix_transfer =
+        Int63.Syntax.(
+          after_suffix_end_offset
+          - t.stats.before_suffix_end_offset
+          - List.fold_left ( + ) Int63.zero worker.suffix_transfers);
+    }
+end
+
+module Worker_stats : sig
+  type t
+
+  val create : string -> t
+  val incr_objects_traversed : t -> t
+  val add_suffix_transfer : t -> int63 -> t
+  val add_file_size : t -> string -> int63 -> t
+  val finish_current_step : t -> string -> t
+  val finalise : t -> Stats.Latest_gc.worker
+end = struct
+  module S = Stats.Latest_gc
+
+  type t = {
+    stats : S.worker;
+    current_stepname : string;
+    prev_wtime : float;
+    prev_stime : float;
+    prev_utime : float;
+    prev_rusage : S.rusage;
+    prev_ocaml_gc : S.ocaml_gc;
+  }
+  (** [t] is the running state while computing the stats *)
+
+  let is_darwin =
+    lazy
+      (try
+         match Unix.open_process_in "uname" |> input_line with
+         | "Darwin" -> true
+         | _ -> false
+       with Unix.Unix_error _ -> false)
+
+  let get_wtime () =
+    (Mtime_clock.now () |> Mtime.to_uint64_ns |> Int64.to_float) /. 1e9
+
+  let get_stime () = Rusage.((get Self).stime)
+  let get_utime () = Rusage.((get Self).utime)
+
+  let get_rusage : unit -> S.rusage =
+   fun () ->
+    let Rusage.{ maxrss; minflt; majflt; inblock; oublock; nvcsw; nivcsw; _ } =
+      Rusage.(get Self)
+    in
+    let maxrss =
+      if Lazy.force is_darwin then Int64.div maxrss 1000L else maxrss
+    in
+    S.{ maxrss; minflt; majflt; inblock; oublock; nvcsw; nivcsw }
+
+  let get_ocaml_gc : unit -> S.ocaml_gc =
+   fun () ->
+    let open Stdlib.Gc in
+    let v = quick_stat () in
+    S.
+      {
+        minor_words = v.minor_words;
+        promoted_words = v.promoted_words;
+        major_words = v.major_words;
+        minor_collections = v.minor_collections;
+        major_collections = v.major_collections;
+        heap_words = v.heap_words;
+        compactions = v.compactions;
+        top_heap_words = v.top_heap_words;
+        stack_size = v.stack_size;
+      }
+
+  let create : string -> t =
+   fun first_stepname ->
+    (* Reseting all irmin-pack stats. We'll reset again at every step. *)
```

Comment:
Maybe add a note that this is okay since we are in the worker process and therefore will not mess up the main process stats.